### PR TITLE
[MUGEN] Add latent_shape to VideoEncoder API

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #251
* #237
* #236
* #235

## Summary
Adds method to `VideoEncoder` that calculates theoretical latent shape based on number of downsampling layers, stride. Add unit tests for this method

## Test plan
`pytest test/models/test_video_vqvae.py -vv`
```
============================================================================ test session starts =============================================================================
platform linux -- Python 3.9.12, pytest-7.1.1, pluggy-1.0.0 -- /fsx/users/rafiayub/conda/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /data/home/rafiayub/torchmultimodal, configfile: pyproject.toml
plugins: hydra-core-1.1.2, cov-3.0.0, mock-3.8.2
collected 10 items                                                                                                                                                           

test/models/test_video_vqvae.py::TestAttentionResidualBlock::test_hidden_dim_assertion PASSED                                                                          [ 10%]
test/models/test_video_vqvae.py::TestAttentionResidualBlock::test_forward PASSED                                                                                       [ 20%]
test/models/test_video_vqvae.py::TestVideoEncoder::test_forward PASSED                                                                                                 [ 30%]
test/models/test_video_vqvae.py::TestVideoEncoder::test_latent_shape PASSED                                                                                            [ 40%]
test/models/test_video_vqvae.py::TestVideoDecoder::test_forward PASSED                                                                                                 [ 50%]
test/models/test_video_vqvae.py::TestVideoVQVAE::test_encode PASSED                                                                                                    [ 60%]
test/models/test_video_vqvae.py::TestVideoVQVAE::test_decode PASSED                                                                                                    [ 70%]
test/models/test_video_vqvae.py::TestVideoVQVAE::test_tokenize PASSED                                                                                                  [ 80%]
test/models/test_video_vqvae.py::TestVideoVQVAE::test_forward PASSED                                                                                                   [ 90%]
test/models/test_video_vqvae.py::test_preprocess_int_conv_params PASSED                                                                                                [100%]

============================================================================= 10 passed in 1.51s =============================================================================
```

Differential Revision: [D38524576](https://our.internmc.facebook.com/intern/diff/D38524576)

### CODE REVIEW

File: torchmultimodal/models/video_vqvae.py [hunk 1]
```diff
@@ -251,6 +251,17 @@ def __init__(
             attn_hidden_dim, output_dim, kernel_size=1, stride=1
         )
 
+    def get_latent_shape(self, input_shape: Union[Tuple, Size]) -> Tuple:
+        """Return shape of encoder output based on number of downsampling conv layers"""
+        latent_shape = list(input_shape)
+        for layer in self.convs:  # ignore conv_out since it has a stride of 1
+            if isinstance(layer, SamePadConv3d):
+                # SamePadConv should downsample input shape by factor of stride
+                for dim in range(len(input_shape)):
```
Comment (langong347) [inline torchmultimodal/models/video_vqvae.py:260]: turn this into a one liner list comprehension?

File: test/models/test_video_vqvae.py [hunk 2]
```diff
@@ -87,6 +87,25 @@ def encoder(self, params):
         enc.eval()
         return enc
 
+    @pytest.fixture
+    def downsampler(self, params):
+        in_channel_dims, _, kernel_sizes, _ = params
+        strides = ((2, 2, 2), (2, 2, 2))
```
Comment (langong347) [inline test/models/test_video_vqvae.py:93]: Use uneven strides, e.g., `((2, 2, 2), (1, 2, 2))`?

File: test/models/test_video_vqvae.py [hunk 3]
```diff
@@ -87,6 +87,25 @@ def encoder(self, params):
...
+            output_dim=2,
+            n_res_layers=1,
+            attn_hidden_dim=2,
+        )
+        model.eval()
+        return model
+
+    @pytest.fixture
+    def big_input(self):
+        return torch.ones(1, 2, 8, 8, 8)
```
Comment (langong347) [inline test/models/test_video_vqvae.py:107]: It's typical for video shape to be different along the dims.
Use uneven values for each dim, e.g., `(1, 2, 8, 64, 64)`?

File: test/models/test_video_vqvae.py [hunk 4]
```diff
@@ -87,6 +87,25 @@ def encoder(self, params):
         enc.eval()
         return enc
 
+    @pytest.fixture
+    def downsampler(self, params):
```
Comment (langong347) [inline test/models/test_video_vqvae.py:91]: Consider creating a single fixture to return a partial encoder instance with `params` exposed as the argument?

File: test/models/test_video_vqvae.py [hunk 5]
```diff
@@ -105,6 +124,17 @@ def test_forward(self, input_tensor, encoder):
         )
         assert_expected(actual, expected, rtol=0, atol=1e-4)
 
+    def test_get_latent_shape(self, downsampler, big_input):
```
Comment (langong347) [inline test/models/test_video_vqvae.py:127]: We could also combine `test_get_latent_shape` and `test_encoder_latent_shape` into one test with `actual` being the output of the forward pass and `expected` being that of `get_latent_shape`? Because ultimately we are more interested in the equality of the two.

**Review by langong347 — APPROVED** [review]
LGTM. Please address linter complaint.

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38524576).
